dataflow/flex-templates/streaming_beam/streaming_beam.py

```diff
@@ -0,0 +1,149 @@
+# limitations under the License.
+
+"""An Apache Beam streaming pipeline example.
+
+It reads JSON encoded messages from Pub/Sub, transforms the message data and
+writes the results to BigQuery.
+"""
+
+from __future__ import absolute_import
+
+import argparse
+import json
+import logging
+import time
+
+import apache_beam as beam
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.options.pipeline_options import SetupOptions
+from apache_beam.options.pipeline_options import StandardOptions
+import apache_beam.transforms.window as window
+
+
+# Defines the BigQuery schema for the output table.
+SCHEMA = ('url:STRING,'
+          'num_reviews:INTEGER,'
+          'score:FLOAT64,'
+          'first_date:TIMESTAMP,'
+          'last_date:TIMESTAMP')
+
+
+class ParseJsonMessageDoFn(beam.DoFn):
+  """Parse the json encoded text message into a dict."""
+
+  def __init__(self):
+    beam.DoFn.__init__(self)
+
+  def process(self, element):
+    """Returns an iterator over the tuple: (message's url, dict).
+
+    Args:
+      element: the element json message to be processed.
+    Returns:
+      The tuple of url string and the dict.
+      The dict contain the keys: 'url', 'review', 'score' and 'processing_time'.
+    """
+    row = json.loads(element)
+    row['score'] = 0.0
+    if row['review'] == 'positive':
+      row['score'] = 1.0
+    row['processing_time'] = int(time.time())
+    return [(row['url'], row)]
+
+
+def run(argv=None, save_main_session=True):
+  """Build and run the pipeline."""
+  parser = argparse.ArgumentParser()
+  parser.add_argument(
+      '--output_table',
+      help=(
+          'Output BigQuery table for results specified as: '
+          'PROJECT:DATASET.TABLE or DATASET.TABLE.'))
+
+  group = parser.add_mutually_exclusive_group(required=True)
+  group.add_argument(
+      '--input_topic',
+      help=(
+          'Input PubSub topic of the form '
+          '"projects/<PROJECT>/topics/<TOPIC>".'))
+  group.add_argument(
+      '--input_subscription',
+      help=(
+          'Input PubSub subscription of the form '
+          '"projects/<PROJECT>/subscriptions/<SUBSCRIPTION>."'))
+  known_args, pipeline_args = parser.parse_known_args(argv)
+
+  pipeline_options = PipelineOptions(pipeline_args)
+  pipeline_options.view_as(SetupOptions).save_main_session = save_main_session
+  pipeline_options.view_as(StandardOptions).streaming = True
+  with beam.Pipeline(options=pipeline_options) as p:
+
+    # Read from PubSub into a PCollection.
+    if known_args.input_subscription:
+      messages = (
+          p
+          | beam.io.ReadFromPubSub(subscription=known_args.input_subscription).
+          with_output_types(bytes))
+    else:
+      messages = (
+          p
+          | beam.io.ReadFromPubSub(
+              topic=known_args.input_topic).with_output_types(bytes))
+
+    lines = messages | 'decode' >> beam.Map(lambda x: x.decode('utf-8'))
```

Comment:
I would chain this with the read from subscription transform to get rid of this temporary `lines` variable.

For the transform name, can we have something more explicit?
`'UTF-8 bytes to string'` maybe

Also, we should rename `x` to something like `message` or `msg`.

Reply:
Done.